cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java

```diff
@@ -128,10 +126,63 @@ public Collection<StructuredRow> multiRead(Collection<? extends Collection<Field
   @Override
   public CloseableIterator<StructuredRow> scan(Range keyRange, int limit) throws InvalidFieldException {
     LOG.trace("Table {}: Scan range {} with limit {}", schema.getTableId(), keyRange, limit);
-    return new LimitIterator(Collections.singleton(new ScannerIterator(getScanner(keyRange), schema)).iterator(),
-                             limit);
+    return new LimitIterator(Collections.singleton(getFilterByRangeIterator(keyRange)).iterator(), limit);
   }
 
+  /**
+   * Scan the nosql db with a key range. It gets the longest prefix keys,
+   * scan the DB and do a post filtering on the result.
+   *
+   * @param keyRange key range to scan
+   * @return the iterator after filtering
+   * @throws InvalidFieldException if the key is not primary key
```

Comment:
I don't think key must be fully primary key. It should have a prefix, but not full key

Reply:
you're right, this will throw `InvalidFieldException` if the key range contains a field that is not a primary key. It's required in nosql implementaions

Reply:
OK, so basically you only allow scans by primary key but without all fields.

Reply:
yes correct